File: dqoai/src/main/java/ai/dqo/connectors/sqlserver/SqlServerParametersSpec.java
/*
 * Copyright © 2021 DQO.ai (support@dqo.ai)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ai.dqo.connectors.sqlserver;

import ai.dqo.connectors.ConnectionProviderSpecificParameters;
import ai.dqo.core.secrets.SecretValueProvider;
import ai.dqo.metadata.id.ChildHierarchyNodeFieldMap;
import ai.dqo.metadata.id.ChildHierarchyNodeFieldMapImpl;
import ai.dqo.metadata.sources.BaseProviderParametersSpec;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.EqualsAndHashCode;
import picocli.CommandLine;

import java.util.Collections;
import java.util.Map;
import java.util.Objects;

/**
 * Microsoft SQL Server connection parameters.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
public class SqlServerParametersSpec extends BaseProviderParametersSpec
        implements ConnectionProviderSpecificParameters {
    private static final ChildHierarchyNodeFieldMapImpl<SqlServerParametersSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(BaseProviderParametersSpec.FIELDS) {
        {
        }
    };

    @CommandLine.Option(names = {"--sqlserver-host"}, description = "SQL Server host name")
    @JsonPropertyDescription("SQL Server host name. Supports also a ${SQLSERVER_HOST} configuration with a custom environment variable.")
    private String host;

    @CommandLine.Option(names = {"--sqlserver-port"}, description = "SQL Server port number", defaultValue = "1433")
    @JsonPropertyDescription("SQL Server port name. The default port is 1433. Supports also a ${SQLSERVER_PORT} configuration with a custom environment variable.")
    private String port;

    @CommandLine.Option(names = {"--sqlserver-database"}, description = "SQL Server database name. The value could be in the format ${ENVIRONMENT_VARIABLE_NAME} to use dynamic substitution.")
    @JsonPropertyDescription("SQL Server database name. The value could be in the format ${ENVIRONMENT_VARIABLE_NAME} to use dynamic substitution.")
    private String database;

    @CommandLine.Option(names = {"--sqlserver-user"}, description = "SQL Server user name. The value could be in the format ${ENVIRONMENT_VARIABLE_NAME} to use dynamic substitution.")
    @JsonPropertyDescription("SQL Server user name. The value could be in the format ${ENVIRONMENT_VARIABLE_NAME} to use dynamic substitution.")
    private String user;

    @CommandLine.Option(names = {"--sqlserver-password"}, description = "SQL Server database password. The value could be in the format ${ENVIRONMENT_VARIABLE_NAME} to use dynamic substitution.")
    @JsonPropertyDescription("SQL Server database password. The value could be in the format ${ENVIRONMENT_VARIABLE_NAME} to use dynamic substitution.")
    private String password;

    @CommandLine.Option(names = {"--sqlserver-options"}, description = "SQL Server connection 'options' initialization parameter. For example setting this to -c statement_timeout=5min would set the statement timeout parameter for this session to 5 minutes.")
    @JsonPropertyDescription("SQL Server connection 'options' initialization parameter. For example setting this to -c statement_timeout=5min would set the statement timeout parameter for this session to 5 minutes. Supports also a ${SQLSERVER_OPTIONS} configuration with a custom environment variable.")
    private String options;

    @CommandLine.Option(names = {"--sqlserver-ssl"}, description = "Connect to SQL Server using SSL", defaultValue = "false")
    @JsonPropertyDescription("Connect to SQL Server using SSL. The default value is false.")
    private Boolean ssl;

    @CommandLine.Option(names = {"--sqlserver-properties"}, description = "SQL Server additional properties that are added to the JDBC connection string")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Map<String, String> properties;

    /**
     * Returns the host name.
     * @return Host name.
     */
    public String getHost() {
        return host;
    }

    /**
     * Sets the host name.
     * @param host New host name.
     */
    public void setHost(String host) {
        setDirtyIf(!Objects.equals(this.host, host));
        this.host = host;
    }

    /**
     * Returns the port name. The value should store an environment variable expression or a numeric MS SQL Server port name.
     * @return Port name or an expression to be extracted.
     */
    public String getPort() {
        return port;
    }

    /**
     * Sets the port name.
     * @param port Port name.
     */
    public void setPort(String port) {
        setDirtyIf(!Objects.equals(this.port, port));
        this.port = port;
    }

    /**
     * Returns a physical database name.
     * @return Physical database name.
     */
    public String getDatabase() {
        return database;
    }

    /**
     * Sets a physical database name.
     * @param database Physical database name.
     */
    public void setDatabase(String database) {
        setDirtyIf(!Objects.equals(this.database, database));
        this.database = database;
    }

    /**
     * Returns the user that is used to log in to the data source (JDBC user or similar).
     * @return User name.
     */
    public String getUser() {
        return user;
    }

    /**
     * Sets a user name.
     * @param user User name.
     */
    public void setUser(String user) {
        setDirtyIf(!Objects.equals(this.user, user));
        this.user = user;
    }

    /**
     * Returns a password used to authenticate to the server.
     * @return Password.
     */
    public String getPassword() {
        return password;
    }

    /**
     * Sets a password that is used to connect to the database.
     * @param password Password.
     */
    public void setPassword(String password) {
        setDirtyIf(!Objects.equals(this.password, password));
        this.password = password;
    }

    /**
     * Returns the custom connection initialization options.
     * @return Connection initialization options.
     */
    public String getOptions() {
        return options;
    }

    /**
     * Sets the connection initialization options.
     * @param options Connection initialization options.
     */
    public void setOptions(String options) {
        setDirtyIf(!Objects.equals(this.options, options));
        this.options = options;
    }

    /**
     * Returns the flag to require SSL connection.
     * @return True - require an SSL connection.
     */
    public Boolean getSsl() {
        return ssl;
    }

    /**
     * Sets a flag to require an SSL connection.
     * @param ssl True - ssl connection is required.
     */
    public void setSsl(Boolean ssl) {
        setDirtyIf(!Objects.equals(this.ssl, ssl));
        this.ssl = ssl;
    }

    /**
     * Returns a key/value map of additional properties that are included in the JDBC connection string.
     * @return Key/value dictionary of additional JDBC properties.
     */
    public Map<String, String> getProperties() {
        return properties;
    }

    /**
     * Sets a dictionary of additional connection parameters that are added to the JDBC connection string.
     * @param properties Key/value dictionary with extra parameters.
     */
    public void setProperties(Map<String, String> properties) {
        setDirtyIf(!Objects.equals(this.properties, properties));
        this.properties = properties != null ? Collections.unmodifiableMap(properties) : null;
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Creates and returns a deep copy of this object.
     */
    @Override
    public SqlServerParametersSpec deepClone() {
        SqlServerParametersSpec cloned = (SqlServerParametersSpec) super.deepClone();
        return cloned;
    }

    /**
     * Creates a trimmed and expanded version of the object without unwanted properties, but with all variables like ${ENV_VAR} expanded.
     * @return Trimmed and expanded version of this object.
     */
    public SqlServerParametersSpec expandAndTrim(SecretValueProvider secretValueProvider) {
        SqlServerParametersSpec cloned = this.deepClone();
        cloned.host = secretValueProvider.expandValue(cloned.host);
        cloned.port = secretValueProvider.expandValue(cloned.port);
        cloned.database = secretValueProvider.expandValue(cloned.database);
        cloned.user = secretValueProvider.expandValue(cloned.user);
        cloned.password = secretValueProvider.expandValue(cloned.password);
        cloned.options = secretValueProvider.expandValue(cloned.options);
        cloned.properties = secretValueProvider.expandProperties(cloned.properties);

        return cloned;
    }
}

File: distribution/python/dqops/client/models/sql_server_parameters_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.sql_server_authentication_mode import SqlServerAuthenticationMode
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.sql_server_parameters_spec_properties import (
        SqlServerParametersSpecProperties,
    )


T = TypeVar("T", bound="SqlServerParametersSpec")


@_attrs_define
class SqlServerParametersSpec:
    """
    Attributes:
        host (Union[Unset, str]): SQL Server host name. Supports also a ${SQLSERVER_HOST} configuration with a custom
            environment variable.
        port (Union[Unset, str]): SQL Server port number. The default port is 1433. Supports also a ${SQLSERVER_PORT}
            configuration with a custom environment variable.
        database (Union[Unset, str]): SQL Server database name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME}
            format to use dynamic substitution.
        user (Union[Unset, str]): SQL Server user name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to
            use dynamic substitution.
        password (Union[Unset, str]): SQL Server database password. The value can be in the ${ENVIRONMENT_VARIABLE_NAME}
            format to use dynamic substitution.
        disable_encryption (Union[Unset, bool]): Disable SSL encryption parameter. The default value is false. You may
            need to disable encryption when SQL Server is started in Docker.
        authentication_mode (Union[Unset, SqlServerAuthenticationMode]):
        properties (Union[Unset, SqlServerParametersSpecProperties]): A dictionary of custom JDBC parameters that are
            added to the JDBC connection string, a key/value dictionary.
    """

    host: Union[Unset, str] = UNSET
    port: Union[Unset, str] = UNSET
    database: Union[Unset, str] = UNSET
    user: Union[Unset, str] = UNSET
    password: Union[Unset, str] = UNSET
    disable_encryption: Union[Unset, bool] = UNSET
    authentication_mode: Union[Unset, SqlServerAuthenticationMode] = UNSET
    properties: Union[Unset, "SqlServerParametersSpecProperties"] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        host = self.host
        port = self.port
        database = self.database
        user = self.user
        password = self.password
        disable_encryption = self.disable_encryption
        authentication_mode: Union[Unset, str] = UNSET
        if not isinstance(self.authentication_mode, Unset):
            authentication_mode = self.authentication_mode.value

        properties: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.properties, Unset):
            properties = self.properties.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if host is not UNSET:
            field_dict["host"] = host
        if port is not UNSET:
            field_dict["port"] = port
        if database is not UNSET:
            field_dict["database"] = database
        if user is not UNSET:
            field_dict["user"] = user
        if password is not UNSET:
            field_dict["password"] = password
        if disable_encryption is not UNSET:
            field_dict["disable_encryption"] = disable_encryption
        if authentication_mode is not UNSET:
            field_dict["authentication_mode"] = authentication_mode
        if properties is not UNSET:
            field_dict["properties"] = properties

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.sql_server_parameters_spec_properties import (
            SqlServerParametersSpecProperties,
        )

        d = src_dict.copy()
        host = d.pop("host", UNSET)

        port = d.pop("port", UNSET)

        database = d.pop("database", UNSET)

        user = d.pop("user", UNSET)

        password = d.pop("password", UNSET)

        disable_encryption = d.pop("disable_encryption", UNSET)

        _authentication_mode = d.pop("authentication_mode", UNSET)
        authentication_mode: Union[Unset, SqlServerAuthenticationMode]
        if isinstance(_authentication_mode, Unset):
            authentication_mode = UNSET
        else:
            authentication_mode = SqlServerAuthenticationMode(_authentication_mode)

        _properties = d.pop("properties", UNSET)
        properties: Union[Unset, SqlServerParametersSpecProperties]
        if isinstance(_properties, Unset):
            properties = UNSET
        else:
            properties = SqlServerParametersSpecProperties.from_dict(_properties)

        sql_server_parameters_spec = cls(
            host=host,
            port=port,
            database=database,
            user=user,
            password=password,
            disable_encryption=disable_encryption,
            authentication_mode=authentication_mode,
            properties=properties,
        )

        sql_server_parameters_spec.additional_properties = d
        return sql_server_parameters_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
